src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs

```diff
@@ -128,7 +140,13 @@ protected void TryStart()
                 }
             }
 
-            OnReadStarted();
+            Task readTask = OnReadStartedAsync();
+            if (!readTask.IsCompletedSuccessfully)
+            {
+                return readTask;
+            }
+
+            return Task.CompletedTask;
```

Comment:
Why not:
```suggestion
            return OnReadStartedAsync();
```
If the OnReadStartedAsync task is complete then return just return it. And if it is not complete then also just return it...